Actors revision 2 (#1321)
apple/swift-evolution: proposals/0306-actors.md

}
```

A closure formed within an actor-isolated context is actor-isolated if it is non-`@Sendable`, and non-isolated if it is `@Sendable`. For the examples above:

> * The closure passed to `Rask.runDetached` is non-isolated because that function requires a `@Sendable` function to be passed to it.

Reviewer: ```suggestion
* The closure passed to `detach` is non-isolated because that function requires a `@Sendable` function to be passed to it.
```

}  
```

> Actors are similar to classes in all respects independent of isolation: actor types can have `static` and `class` methods, properties, and subscripts. All of the attributes that apply to classes apply to actors in much the same way, except where those semantics conflict with actor isolation.

Reviewer: `static` stay though I hope? Just removal of the test here to compare with classes; no reason not to do static IMHO, very useful for IDs and stuff

Author: static is fine, I'll comment on it

Reviewer 2: Might be good to clarify the isolation behavior (or not) of `static var`'s if they are allowed.  It might be worth banning them from actors?

Author: Sure, I can clarify that as well.

Reviewer: Sounds good such properties to ban even, agreed :-)

  
  func useAF(array: [Int]) {
    array.map(self.f)                     // okay

>     detach(operation: self.g)   // error: self.g has non-sendable type () -> Double that cannot be converted to a @Sendable function type

Reviewer: ```suggestion
    detach(operation: self.g)             // error: self.g has non-sendable type () -> Double that cannot be converted to a @Sendable function type
```

extension A {
  func useAFDesugared(a: A, array: [Int]) {
    array.map { f($0) } )      // okay

>     detach { g() }   // error: self is non-isolated, so call to `g` cannot be synchronous

Reviewer: ```suggestion
    detach { g() }             // error: self is non-isolated, so call to `g` cannot be synchronous
```

## Alternatives considered

### Actor inheritance

Earlier pitches and the first reviewed version of this proposal allowed actor inheritance. Actor inheritance followed the rules of class inheritance, albeit with specific additional rules required to maintain actor isolation:

* An actor could not inherit from a class, and vice-versa.
* An overriding declaration must not be more isolated than the overridden declaration.

> Subsequent review discussion determined that the conceptual cost of actor inheritance outweighed its usefulness, so it has been removed from this proposal. The form that actor inheritance would take in the language is well-understand from prior iterations of this proposal and its implementation, so this feature could be re-introduced at a later time.

Reviewer: 👍 I warmed up to this as well once I realized the hell that class style initializers are 🤣+😭

> ### Cross-actor lets

Reviewer: I'm confused what this section is claiming; 

Are we removing this "very permissiveness" from local actors or not by default or keeping it and this future work explains... I'm not sure what?
 
--- 

The permissive synchronous let access is definitely a bit "weird" and "not pure actor model" etc. but I also don't think it's a huge deal to do so. As long as the type allowed for such access is required to be `Sendable`. 

Distributed actors obviously will not allow such synchronous access either way, _even if_ `nonisolated let x: X` so I don't think it matters either way for distributed actors and the difference will exist anyway. Though one could reach into `distributed actor DA { nonisolated let x: X }` with `da.whenLocal { $0.x }` without having to do any extra "with unsafe actor state" dance etc. Long story short: distributed actors don't matter much for allowing cross-actor lets or not, because in all shapes (even non isolated) it's banned thing.

Author: We will call this out for discussion in the second review
